docs: async-hooks documentation

thlorenz opened the pull request:
WIP PR to get Feedback.

@rvagg @fishrock123 @trevnorris

thlorenz committed:
docs: async-hooks documentation

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,275 @@
+all optional. Returns an `AsyncHook`
+
+
+* `init` {function}
```
nit: the value inside of the `{}` is usually capitalized in core docs at least.

thlorenz replied:
cool will fix

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,275 @@
+
+
+function init(uid, type, triggerId, handle) {
+  const activity = { uid,  type, parentUid }
```
`parentUid -> triggerId`

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,275 @@
+process.on('exit', onexit)
+function onexit() {
+  // Stop listening on the above hooks
+  asyncHook.disable()
```
thought: should we automatically stop listening in the `'exit'` event?

thlorenz replied:
I'd opt for not doing any magic here. It doesn't matter much anyways since the process is exiting ...

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,275 @@
+    * `uid` {number} a unique id for the async operation
+
+
+### AsyncHook.enable
```
This should either be `asyncHook.enable()` or `AsyncHook#enable()`. Like this it looks like a static member (i.e. not one on the prototype).

thlorenz replied:
which one of these two is more common given this is a module method? cc @fishrock123

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,275 @@
+### async_hooks.currentId()
+
+
+Returns the uid of the current context.
```
nit: "current _execution_ context". remove a hair more ambiguity.

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,275 @@
+embedder API.
+
+
+## Embedder API
```
The preferred embedder API is to use `AsyncEvent`. The `async_hooks.emitInit()`, etc., were made for internals, but I could find reasonable need for them in user code. But want to make clear they should attempt to use `AsyncEvent` first, if possible.

thlorenz replied:
OK will dig into that a bit more `AsyncEvent` and update the docs as far as I'll understand things at that point ...

trevnorris reviewed (commented):
Docs will need to be word wrapped before going into core. I'll let you decide when to take care of that.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,275 @@
+
+
+The `async-hooks` module provides an API to register callbacks tracking the lifetime of asynchronous handles created
+inside a Node.js application.
```
Needs a terminology section: https://github.com/nodejs/node-eps/blob/e35d4813fdbbd99a751296e24361dba0d0dd9e10/XXX-asyncwrap-api.md#terminology

The top-level description should probably not include the word "handle".

Fishrock123 reviewed (changes requested):
So, al the Node.js docs follow some order:

All sections and content within them must be alphabetical, the exception being if you have API and subheadings in the same spot, then the API references must be alphabetical but the sub-heading should usually go last.

Also, we tend to put examples _after_ API references, except in the case of a top-level example.

thlorenz commented:
@Fishrock123 @trevnorris, ready for another review. Fixed a bunch of nits and integrated big parts from the [EPS](https://github.com/nodejs/node-eps/blob/async-wrap-ep/XXX-asyncwrap-api.md)

Most now comes from this EPS, except I reordered things a bit to clearly separate the different parts of the API.
I also removed some parts that described implementation details not useful to the users.

I also removed the original example I had in there, as now lots of examples are already included in numerous places.

LMK what you think at this point.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+
+
+> Stability: 2 - Stable
+
```
extra spaces?

thlorenz replied:
could you clarify? Missing spaces, need to add them and where?

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+Requests are short lived data structures created to accomplish one task. The
+callback for a request should always and only ever fire one time. Which is when
+the assigned task has either completed or encountered an error. Requests are
+used by handles to perform these tasks. Such as accepting a new connection or
```
`tasks, such as` -- the last sentence probably shouldn't be it's own.

thlorenz replied:
OK will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+const async_hooks = require('async_hooks');
+
+// Return the id of the current execution context. Useful for tracking state
+// and retrieving the handle of the current trigger without needing to use an
```
You mean retrieving the handle if the user has stored it?

thlorenz replied:
Should probably be `id` instead of handle .. I agree it's confusing.
This came from the EP (slightly adapted) so @trevnorris should know.

trevnorris replied:
> ... retrieving the handle of the current trigger

That comment dates back to when I had implemented a map to keep track of all resources, and there was `async_hooks.getResourceById()`. Which would return the resource from the map, but it ended up being too costly. So it was removed.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+// current execution scope to fire.
+const tid = aysnc_hooks.triggerId();
+
+// Create a new instance of AsyncHook. All of these callbacks are optional.
```
`a new AsyncHook instance.`

thlorenz replied:
Hairsplitting, but I've got no strong opinion, so will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+
+// Disable listening for new asynchronous events. Though this will not prevent
+// callbacks from firing on asynchronous chains that have already run within
+// the scope of an enabled AsyncHook instance.
```
It will actually prevent those

thlorenz replied:
Need clarification from @trevnorris here.

trevnorris replied:
I updated the EP to simply say:

> // Disable listening for all new asynchronous events.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+//
+
+// init() is called during object construction. The resource may not have
+// completed construction when this callback runs. So all fields of the
```
`runs, so ...`

thlorenz replied:
I'd rather change `So` to `Therefore` as that clearer conveys the meaning IMO.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+function before(id) { }
+
+// after() is called just after the resource's callback has finished, and will
+// always fire. Except in the case of an uncaught exception.
```
`has finished. This hook will not be called if an uncaught exception occurred.`

thlorenz replied:
Will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+
+Registers functions to be called for different lifetime events of each async
+operation.
+The above callbacks registered via an `AsyncHooks` instance fire during specific
```
remove `above`?

thlorenz replied:
It's referring to the above section, but could remove it.

trevnorris replied:
If that is removed I'd suggest explicitly mentioning all four callbacks. e.g.

> The callbacks `init()`/`before()`/`after()`/`destroy()` are registered via...

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+calls centers around the lifetime of the `AsyncWrap` C++ class. These callbacks
+will also be called to emulate the lifetime of handles and requests that do not
+fit this model. For example, `HTTPParser` instances are recycled to improve
+performance. So the `destroy()` callback would be called manually after a
```
`is called manually`

thlorenz replied:
Will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+**Error Handling**: If any `AsyncHook` callbacks throw, the application will
+print the stack trace and exit. The exit path does follow that of any uncaught
+exception, except for the fact that it is not catchable by an uncaught
+exception handler, so any `'exit'` callbacks will fire. Unless the application
```
maybe remove `, except for the fact that it is not catchable by an uncaught exception handler`

and edit `, however `'exit'` callbacks will still fire.`

thlorenz replied:
Agreed, much simpler and we don't need those implementation details.
Will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+be printed and the application will exit, leaving a core file.
+
+The reason for this error handling behavior is that these callbacks are running
+at potentially volatile points in an object's lifetime. For example during
```
`For example,`

thlorenz replied:
Will fix

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+The reason for this error handling behavior is that these callbacks are running
+at potentially volatile points in an object's lifetime. For example during
+class construction and destruction. Because of this, it is deemed necessary to
+bring down the process quickly as to prevent an unintentional abort in the
```
remove `as`

thlorenz replied:
Will change `as` to `in order` - sound ok?

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+#### Hook Callbacks
+
+Key events in the lifetime of asynchronous events have been categorized into
+four areas. On instantiation, before/after the callback is called and when the
```
`four areas: instantiation ...`

`is called, and`

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+
+Every new resource is assigned a unique id.
+
+The `type` is a String that represents the type of resource that caused
```
Should link to the list here?

thlorenz replied:
Definitely, just don't know where that list exists ATM.
Ideally it'd be linking to a place that gets updated in case that list changes.

trevnorris replied:
Currently it only exists in source code. I expect this list will be maintained in the documentation. The list in `src/async-wrap.h`.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+to define their own `type` when using the public embedder API.
+
+**Note:** It's possible to have name collisions. So embedders are recommended
+to use a prefix of sorts to prevent this.
```
`It is possible to have type name collisions. Embedders are recommended to use unique prefixes per module to prevent collisions when listening to the hooks.`

thlorenz replied:
Much better, thanks. Will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+const async_hooks = require('async_hooks');
+
+asyns_hooks.createHook({
+  init: (id, type, triggerId) => {
```
This can be `init(id, type, triggerId) {` ... Object literals syntax :)

ditto for other examples

thlorenz replied:
ah yeah, wasn't even aware you can do that outside of classes.
I never inline define functions inside an object literal like this.
Not a good practice in general and ChromeDevtools still has problems debugging these i.e. hard to get a break point working.
Ran into this recently with @trevnorris .

So do you guys think we should change this in general to either use a proper class definition or somehow un-inline the function and move the declaration to the top to lead by example with good practices?

Other wise I'll just remove the `=>`s where possible.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+No where here do we see the `TCPWRAP` created; which was the reason for
+`console.log()` being called. This is because binding to a port without a
+hostname is actually synchronous, but to maintain a completely asynchronous API
+the user's callback is placed in a `process.nextTick()`.
```
The language should generally be moved from conversational ("we") to declarative ("it").

thlorenz replied:
Agreed, this isn't a blog post :) . Will fix by rewording it.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+
+When an asynchronous operation is triggered (such as a TCP server receiving a
+new connection) or completes (such as writing data to disk) a callback is
+called to notify node. The `before()` callback is called just before said
```
`Node`

thlorenz replied:
Will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+
+Called immediately after the callback specified in `before()` is completed. If
+an uncaught exception occurs during execution of the callback then `after()`
+will run after `'uncaughtException'` or a `domain`'s handler runs.
```
```
the `'uncaughtException'` event
```

thlorenz replied:
Will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+**Note:** Some resources depend on GC for cleanup. So if a reference is made to
+the `resource` object passed to `init()` it's possible that `destroy()` is
+never called. Causing a memory leak in the application. Of course if you know
+the resource doesn't depend on GC then this isn't an issue.
```
We should have a list of such resources here.

thlorenz replied:
A list of all possible resources like you requested to be linked earlier?
Or a list of resources that could cause a leak if held on to?

trevnorris replied:
It's anything that uses `MakeWeak()` in C++. Can throw that list together if needed.

Fishrock123 commented on doc/api/async_hooks.md:
````diff
@@ -0,0 +1,671 @@
+}):
+```
+
+It is important to note that the id returned has to do with execution timing.
````
```
returned from `currentId()`
```

thlorenz replied:
Will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+
+The class `AsyncEvent` was designed to be extended from for embedder's async
+resources. Using this users can easily trigger the lifetime events of their
+own resources.
```
Perhaps we want to put in a thing about that I/O pools need to use this?

thlorenz replied:
Need more details on that or maybe you or @trevnorris can suggest the addition here?

trevnorris replied:
@Fishrock123 can you clarify what you mean?

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+
+// Call after() hooks. It is important that before/after calls are unwound
+// in the same order they are called. Otherwise an unrecoverable exception
+// will be made.
```
Perhaps this comment should go under the actual api header with an example?

thlorenz replied:
Yeah, makes sense. Will fix.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+#### `AsyncEvent(type[, triggerId])`
+
+* arguments
+  * type {String} the type of ascync event
```
`async`

thlorenz replied:
Will fix.

trevnorris replied:
The argument names are supposed to be surrounded in backticks. e.g. `` `type` ``. Please change throughout the document.

Fishrock123 commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+If the user's callback throws an exception then `emitAfter()` will
+automatically be called for all `id`'s on the stack if the error is handled by
+a domain or `'uncaughtException'` handler. So there is no need to guard against
+this.
```
Perhaps we should say it should not be called or an example given?

thlorenz replied:
Not fully sure what you mean. What _should not be called_?

Fishrock123 reviewed (commented):
I think it would also be good to have a section explaining that you must sue `process._rawDebug()` to log within a look so as to prevent an infinite loop.

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+# Async Hooks
+
+
+> Stability: 2 - Stable
```
I believe the initial release will be labeled `1 - Experimental` so that we have a major version to make any necessary tweaks.

thlorenz replied:
Fixed

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+lifetime of asynchronous resources created inside a Node.js application.
+
+At its simplest form it can track metadata about each resource and log it when
+the program exits.
```
This doesn't make much sense to me. The absolute simplest would be to log metadata about asynchronous resources when they occur.

thlorenz replied:
Removed the paragraph

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+async_hooks.createHook({
+  init: (id, type, triggerId) => {
+    const cId = async_hooks.currentId();
+    process._rawDebug(' '.repeat(ws) +
```
Did you add that `console.*()` should never be used from hook callbacks? May need to mention this.

trevnorris commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,671 @@
+    SIGNALWRAP(7): trigger: 4 scope: 4
+>>> 4
+  after:   4
+after:   5
```
The script above, and this output, were both updated in the EP. Please reference.

thlorenz replied:
Updated the output .. didn't see any changes to the script.
Note that we're using class like function declarations (instead of arrow functions) as per request by @Fishrock123

trevnorris reviewed (changes requested):
Thanks for the work. I like the added example of using the `AsyncEvent` API.

thlorenz commented:
@Fishrock123 @trevnorris addressed the nits.

Still need input/clarification on the below

- [linking to resource list](https://github.com/thlorenz/node/pull/2#discussion_r98048067)
- [waiting for @trevnorris to provide MakeWeak list](https://github.com/thlorenz/node/pull/2#discussion_r98049722)
- [need clarification from @Fishrock123](https://github.com/thlorenz/node/pull/2#discussion_r98050085)
- [also need clarification from @Fishrock123](https://github.com/thlorenz/node/pull/2#discussion_r98050506)

thlorenz committed:
docs: addressing nits pointed out by Trevor

thlorenz commented:
So I fixed the nits pointed out by Trevor, but still need some clarification from @Fishrock123 for some of the above.

brycebaril commented on doc/api/async_hooks.md:
````diff
@@ -0,0 +1,670 @@
+TTYWRAP(6) -> Timeout(4) -> TIMERWRAP(5) -> TickObject(3) -> root(1)
+```
+
+The `TCPWRAP` isn't part of this graph; evne though it was the reason for
````
s/evne/even/

brycebaril commented on doc/api/async_hooks.md:
````diff
@@ -0,0 +1,670 @@
+```js
+const async_hooks = require('async_hooks');
+
+// Return the id of the current execution context.
````
Probably worth defining 'execution context' somewhere.

brycebaril commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,670 @@
+hostname is actually synchronous, but to maintain a completely asynchronous API
+the user's callback is placed in a `process.nextTick()`.
+
+The graph only shows **when** a resource was created. Not **why**. So to track
```
>Not **why**. So to track the **why** use `triggerId`.

This is a bit awkward

brycebaril commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,670 @@
+    event
+* Returns {AsyncEvent} A reference to `asyncHook`.
+
+Example usage:
```
This seems less like an embedder's use but an APM/wrapper's use. I'd think that the embedder would put the `emitBefore and `emitAfter` inside `function this.db.get` not the `getInfo` wrapper?

thlorenz commented:
@brycebaril let's continue discussion here: https://github.com/nodejs/node/pull/12953

drewfish commented on doc/api/async_hooks.md:
```diff
@@ -0,0 +1,670 @@
+const async_hooks = require('async_hooks');
+
+// Return new unique id for a constructing handle.
+const id = async_hooks.newUid();
```
Should this be `newId()`?
